firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+
+
+class Transpose(Tensor):
+    """An abstract SLATE class representing the tranpose of a tensor."""
+
+    def __init__(self, A):
+        """Constructor for the Transpose class.
+
+        :arg A: a SLATE tensor."""
+
+        slate_assert(not isinstance(A, Scalar),
+                     "Cannot take the transpose of a scalar.")
+        self.children = A
+        Tensor.tensor_id += 1
+        super(Transpose, self).__init__(arguments=A.arguments()[::-1],
+                                        coefficients=A.coefficients(),
+                                        integrals=A.get_ufl_integrals())
+
+    @property
+    def operands(self):
+        """Returns the operand of the transpose operator."""
+        return (self.children, )
+
+    def __str__(self, prec=None):
+        """String representation of a transposed tensor."""
+        return "(%s).T" % self.children
+
+    def __repr__(self):
+        """SLATE representation of a transposed tensor."""
+        return "Transpose(%s)" % self.children
+
+
+class UnaryOp(Tensor):
+    """An abstract SLATE class for representing unary operations on a
+    Tensor object."""
+
+    def __init__(self, A):
+        """Constructor for the UnaryOp class.
+
+        :arg tensor: a SLATE tensor."""
+
+        if not isinstance(A, Tensor):
+            expecting_slate_object(Tensor, type(A))
+        self.children = A
+        Tensor.tensor_id += 1
+        super(UnaryOp, self).__init__(arguments=A.arguments(),
+                                      coefficients=A.coefficients(),
+                                      integrals=self.get_uop_integrals(A))
+
+    @classmethod
+    def get_uop_integrals(cls, A):
+        """Returns the integrals of the resulting tensor after the
+        unary operation (Positive or Negative)
+        is performed.
+
+        Method is implemented in subclass."""
+        pass
+
+    @property
+    def operands(self):
+        """Returns the operand of the unary operation."""
+        return (self.children, )
+
+    def __str__(self, prec=None):
+        """String representation of a resulting tensor after a unary
+        operation is performed."""
+        ops = {operator.neg: '-',
+               operator.pos: '+'}
+        if prec is None or self.prec >= prec:
+            pars = lambda x: x
+        else:
+            pars = lambda x: "(%s)" % x
+
+        return pars("%s%s" % (ops[self.op], self.children.__str__(prec=self.prec)))
+
+    def __repr__(self):
+        """SLATE representation of the resulting tensor."""
+        return "%s(%r)" % (type(self).__name__, self.children)
+
+
+class Negative(UnaryOp):
+    """Abstract SLATE class representing the negation of a tensor object."""
+
+    prec = 1
+    op = operator.neg
+
+    @classmethod
+    def get_uop_integrals(cls, A):
+        """Returns the integrals of Negative(A)."""
+
+        integrals = []
+        for it in A.get_ufl_integrals():
+            integrals.append(-it)
+        return tuple(integrals)
```

Comment:
So that seems odd to me.  Because I would have thought instead that you would just generate `-Matrix` in the code.  Rather than `Matrix(-bunch_of_matrices)`.